### PR TITLE
Make penalty increase/decrease configurable

### PR DESCRIPTION
**Pull request by N-Wouda**
This is one of the things we missed in the baseline, and that was very useful to change for us during the EURO meets NeurIPS competition. The previously hardcoded values remain the default, so this should not impact the sequence of solutions found.

TODO:

- [x] Implement increase/decrease parameters
- [x] Implement # iters for penalty management parameter
- [x] Verify tests
- [ ] Benchmark on cluster (100/0.85/1.2 vs 50/0.3/1.3 on 10x the X-instances)

### CODE REVIEW

**Comment by N-Wouda** [comment]
(this is the first in possibly a sequence of PRs that contribute some changes back upstream, since I ran into a number of minor things while working on our VRPTW solver for the EURO meets NeurIPS competition)

**Comment by N-Wouda** [comment]
All tests still pass with this change in place!

File: Program/AlgorithmParameters.h [hunk 1]
```diff
@@ -14,6 +14,8 @@ struct AlgorithmParameters {
 	int nbElite;			// Number of elite individuals
 	int nbClose;			// Number of closest solutions/individuals considered when calculating diversity contribution
 	double targetFeasible;	// Reference proportion for the number of feasible individuals, used for the adaptation of the penalty parameters
+	double penaltyDecrease;
```
Comment (vidalt) [inline Program/AlgorithmParameters.h]: Add a simple comment here to follow standard practice in this file

File: Program/commandline.h [hunk 2]
```diff
@@ -77,6 +77,10 @@ class CommandLine
 					ap.nbClose = atoi(argv[i+1]);
 				else if (std::string(argv[i]) == "-targetFeasible")
```
Comment (vidalt) [inline Program/commandline.h:80]: You may need to increase the value set at line https://github.com/vidalt/HGS-CVRP/blob/cc9c54fdf005dc1576ac34e044a7a55c18d705ec/Program/commandline.h#L45 to make sure we could specify all available parameters without running into an exception

**Comment by N-Wouda** [comment]
@vidalt sure, will do in a minute. Are the CI failures on Windows something I should worry about? It seems to be complaining about missing files.

**Comment by vidalt** [comment]
Thanks for the quick updates and responses!
Just to give a bit more background on these (increase/decrease) parameters for penalty control.
So far, I had not included them strictly speaking as "parameters" since most values did an equal job without much impact on the algorithm performance. Indeed, as long as the coefficients remain reasonable (e.g., between 0.75 and 0.99 for decrease and between 1.01 and 1.5 for increase), the HGS algorithm quickly converges towards reasonable values for the penalty parameters and then oscillates around it. So, adding two parameters for calibration seemed conceptually an overkill and may distract the user. Your pull request revisits this assumption: have you indeed obtained significantly better results on the CVRP (even on certain subsets of the instances) by calibrating these parameters?

**Comment by N-Wouda** [comment]
>Your pull request revisits this assumption: have you indeed obtained significantly better results on the CVRP (even on certain subsets of the instances) by calibrating these parameters?

I mostly opened this PR because it seemed strange to me that these parameters were hardcoded. I do not have anything specifically for the CVRP, but for the VRPTW instances we solved during the EURO/NeurIPS competition, after tuning we found the following settings to be effective:

    double penaltyIncrease = 1.34;    // regular increase if below target feas
    double penaltyDecrease = 0.32;    // regular decrease if above target feas

Superficially, this seems to confirm that the default for the penalty increase of 1.2 is indeed a reasonable value. For the decrease, however, a lower value seemed to be more effective in the setting of that competition.

We tuned these parameters together with the number of iterations before updating the penalties. We halved that number of iterations from 100 to around 50, and there <100 seemed to almost always be a better choice. It could be that that's also influencing our results. But making that number of iterations variable is considerably more work (and might also influence the sequence of solutions found? - not sure), so I have avoided it here.

How would you me like to proceed?

**Comment by vidalt** [comment]
Hello @N-Wouda, thanks a lot for the insights; this is very interesting!
Your calibration seems to indicate that a policy that does fast drops of penalty values (x 0.3) followed by a slower increase (x 1.3) has some advantages in terms of exploration. Conceptually, this means that the penalty will oscillate around a target value with a "saw-blade" type of penalty function: doing some sharp drops (i.e., going further away from feasibility) and then slowly increasing to recover feasibility. This is consistent with previous findings in some papers (i.e., the concept of strategic oscillation by Fred Glover). Can you compare the difference in performance on the CVRP between the proposed calibration (0.3/1.3) and the original one, doing some repeated runs for statistical significance?
Ideally, a more complete (factorial?) analysis involving multiple runs on all X instances with decrease coefficients in the range 0.9 0.8 0.7 0.6 0.5 0.45 0.4 0.35 0.3 0.25 0.2 0.15 0.1 and increase coefficients such as 1.05 1.1 1.15 1.2 1.25 1.3 1.4 1.5 1.6 1.7 1.8 1.9 2.0 would permit to get the complete picture of how the method behaves when varying these coefficients.

**Comment by vidalt** [comment]
Regarding the comment, "We tuned these parameters together with the number of iterations before updating the penalties". Indeed, if we set the increase/decrease coefficients for the penalties as a tunable parameter, it would also make sense to put the penalty update frequency (currently 100, was it 50 in your tests on the VRPTW?) as a configurable parameter.

**Comment by N-Wouda** [comment]
@vidalt: 

>Can you compare the difference in performance on the CVRP between the proposed calibration (0.3/1.3) and the original one, doing some repeated runs for statistical significance?

Yes, I'll run at least '0.3/1.3' at some point during the Christmas break. I'll have a look at whether I can also run the full scale experiment, but that might be a bit tricky since I also have to run my own research and can't afford to trash my fairshare on our computer cluster :).

>Indeed, if we set the increase/decrease coefficients for the penalties as a tunable parameter, it would also make sense to put the penalty update frequency (currently 100, was it 50 in your tests on the VRPTW?) as a configurable parameter.

We used 47 in the final configuration. I will add this parameter to the PR before running the experiments, and then test, say, 0.3/1.3 and 50. 

Expect an update in a few weeks!

**Comment by N-Wouda** [comment]
>Benchmark on cluster (100/0.85/1.2 vs 50/0.3/1.3 on 10x the X-instances)

I'll get this running tomorrow.

File: Program/commandline.h [hunk 3]
```diff
@@ -42,7 +42,7 @@ class CommandLine
 	// Reads the line of command and extracts possible options
 	CommandLine(int argc, char* argv[])
 	{
-		if (argc % 2 != 1 || argc > 27 || argc < 3)
+		if (argc % 2 != 1 || argc > 31 || argc < 3)
```
Comment (vidalt) [inline Program/commandline.h]: after adding "nbIterPenaltyManagement" should it be 33?

File: Program/commandline.h [hunk 4]
```diff
@@ -105,7 +111,10 @@ class CommandLine
 		std::cout << "[-lambda <int>] Number of solutions created before reaching the maximum population size (i.e., generation size). Defaults to 40 " << std::endl;
 		std::cout << "[-nbElite <int>] Number of elite individuals. Defaults to 5                                                                     " << std::endl;
 		std::cout << "[-nbClose <int>] Number of closest solutions/individuals considered when calculating diversity contribution. Defaults to 4      " << std::endl;
-		std::cout << "[-targetFeasible <double>] target ratio of feasible individuals in the last 100 generatied individuals. Defaults to 0.2         " << std::endl;
+		std::cout << "[-nbIterPenaltyManagement <int>] Number of iterations between penalty updates. Defaults to 100                                  " << std::endl;
```
Comment (vidalt) [inline Program/commandline.h:114]: Now that 100 is a parameter should we put a different wording?

File: README.md [hunk 5]
```diff
@@ -83,7 +83,9 @@ Additional Arguments:
 [-lambda <int>] Number of solutions created before reaching the maximum population size (i.e., generation size). Defaults to 40
 [-nbElite <int>] Number of elite individuals. Defaults to 5                                                                    
 [-nbClose <int>] Number of closest solutions/individuals considered when calculating diversity contribution. Defaults to 4     
-[-targetFeasible <double>] target ratio of feasible individuals in the last 100 generatied individuals. Defaults to 0.2
```
Comment (vidalt) [inline README.md:86]: Should we add nbIterPenaltyManagement in the README?

**Comment by N-Wouda** [comment]
@vidalt I have the X-instances running and expect those to complete later today.

<details>
  <summary>Jobscript generator</summary>

```python
import glob
import pathlib
import re

SCRIPT = """#!/bin/bash

#SBATCH --job-name=hgs
#SBATCH --time={script_limit}
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=1G
#SBATCH --out=slurm/%x-%A_%a.out

module purge
module load GCCcore/11.3.0

for seed in {{1..10}}; do
    PARAMS=(
        Instances/CVRP/{data}.vrp
        Results/{data}-$seed-{exp}.sol
        -seed $seed
        -t {time_limit}
        -penaltyIncrease {penaltyIncrease}
        -penaltyDecrease {penaltyDecrease}
        -nbIterPenaltyManagement {nbIterPenaltyManagement}
    );

    ./hgs ${{PARAMS[@]}};
done
"""


def seconds2string(seconds: int) -> str:
    mins, seconds = divmod(seconds, 60)
    hours, mins = divmod(mins, 60)
    return f"{hours:02d}:{mins:02d}:{seconds:02d}"


def main():
    experiments = [
        dict(  # baseline
            penaltyIncrease=1.2,
            penaltyDecrease=0.85,
            nbIterPenaltyManagement=100,
            targetFeasible=0.2,
        ),
        dict(  # as used in EURO/NeurIPS
            penaltyIncrease=1.3,
            penaltyDecrease=0.3,
            nbIterPenaltyManagement=50,
            targetFeasible=0.4,
        ),
        dict(  # as proposed
            penaltyIncrease=1.3,
            penaltyDecrease=0.3,
            nbIterPenaltyManagement=50,
            targetFeasible=0.2,
        ),
    ]

    for exp, level in enumerate(experiments):
        for data_loc in glob.iglob("Instances/CVRP/X*.vrp"):
            data_loc = pathlib.Path(data_loc)
            script_loc = f"Scripts/{data_loc.stem}-{exp}.sh"

            size = int(re.search(r'-n(\d{1,4})-', data_loc.stem).group(1))
            num_clients = size - 1

            time_limit = int(num_clients * 240 / 100)
            script_limit = 10 * (time_limit + 60)  # one minute leeway, per seed

            with open(script_loc, "w") as fh:
                fh.write(SCRIPT.format(
                    data=data_loc.stem,
                    exp=exp,
                    time_limit=time_limit,
                    script_limit=seconds2string(script_limit),
                    **level,
                ))


if __name__ == "__main__":
    main()
```

</details>

File: Program/commandline.h (hunk 3 above)
Comment (N-Wouda) [reply Program/commandline.h]: Yes, thanks for catching that!

File: Program/commandline.h (hunk 4 above)
Comment (N-Wouda) [reply Program/commandline.h:114]: Do you have a suggested alternative? I have already rephrased `targetFeasible`, `penaltyIncrease`, and `penaltyDecrease`, but I'm happy to change them!

**Comment by N-Wouda** [comment]
I ran three different experiments on the X-instances, each instance solved with ten different seeds. 

The first experiment uses the baseline parameter values of 100/0.85/1.2, the second 50/0.3/1.3, and the third 50/0.3/1.3 plus `targetFeasible=0.4` (like we did in EURO/NeurIPS). Results in the table below:

| Experiment | Gap to BKS |
| -- | -- |
| 100/0.85/1.2 | 0.127% |
| 50/0.3/1.3 | 0.158% |
| 50/0.3/1.3/0.4 | 0.205% |

A larger experiment might find a better parameter configuration, but I do not have the computing resources to investigate that right now. @vidalt how would you like to proceed with this PR?

**Comment by vidalt** [comment]
Hello @N-Wouda, 
Thanks a lot for running these tests! Unfortunately, the experiments did not highlight a better configuration for these parameters, but they also seem to demonstrate that their values matter. If I have a bit of time in the future, I will run some more complete experiments on this. For the time being, I will merge this pull request as it adds functionalities that could help some users, is backward compatible, and does not change the current behavior of the code.

**Pull request merged**